flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -157,10 +167,17 @@ public ZooKeeperStateHandleStore(
         checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
         checkNotNull(state, "State");
         final String path = normalizePath(pathInZooKeeper);
-        if (exists(path).isExisting()) {
-            throw new AlreadyExistException(
-                    String.format("ZooKeeper node %s already exists.", path));
+        final Optional<Stat> nodeStat = nodeExistsButMightBeMarkedAsDeleted(path);
+
+        if (nodeStat.isPresent()) {
+            if (NOT_MARKED_FOR_DELETION.test(nodeStat.get())) {
+                throw new AlreadyExistException(
+                        String.format("ZooKeeper node %s already exists.", path));
+            }
+
+            releaseAndTryRemove(path);
```

Comment:
it's only called during the checkpoint creation (where we shouldn't run into the situation that a checkpoint gets deleted because only one leader access a checkpoint at a time) and when a JobGraph is added during a job submission.

So, in this case, this is only added to make the implementation consistent.